fms/triton/pytorch_ops.py

```diff
@@ -0,0 +1,224 @@
+
+    # First count how many tokens go to each expert
+    cnts = torch.zeros(
+        topk_ids.shape[0], num_experts, dtype=topk_ids.dtype, device=topk_ids.device
+    )
+    cnts.scatter_(1, topk_ids, 1)
+    tokens_per_expert = cnts.sum(dim=0)
+
+    # Then pad the amount to the block size of work
+    tokens_per_expert_block_padded = (
+        torch.floor_divide(tokens_per_expert + block_size - 1, block_size) * block_size
+    )
+
+    # Count how many tokens in total are we computing in the GPU
+    cumsum = tokens_per_expert_block_padded.cumsum(0)
+    total_padded_tokens = cumsum[-1]
+
+    # For allocation purposes, compute the worst case scenario for how
+    # many tokens we could be doing work for if the RNG gods are bad
+    max_total_padded_tokens = (
+        (topk_ids.numel() + num_experts * (block_size - 1))
+        if topk_ids.numel() > num_experts
+        else (topk_ids.numel() + 1) * block_size
+    )
+
+    # Compute what MoE expert corresponds to each block of work
+    # A block of work consists of tokens that all go to the same expert
+    # There are total_padded_tokens // block_size blocks of work, but to
+    # simplify kernel launches, we allocate based on worst case and use
+    # max_total_padded_tokens instead.
+    expert_block_mapping = torch.zeros(
+        max((max_total_padded_tokens + block_size - 1) // block_size + 1, num_experts),
+        dtype=topk_ids.dtype,
+        device=topk_ids.device,
+    )
+
+    num_blocks_per_expert = cumsum.div(block_size, rounding_mode="floor")
+    ones = torch.ones_like(expert_block_mapping)
+    expert_block_mapping.scatter_add_(0, num_blocks_per_expert, ones)
+    expert_block_mapping = expert_block_mapping.cumsum(0)
+
+    # Create the mapping between token idxs in the input tensor and
+    # the tokens that will go in each work block
+
+    # Count how many pad tokens need adding to the list of tokens in
+    # topk_ids, then create a tensor that will fill in the right spots
+    # when doing an argsort later to compute padded_token_ids_per_block
+    cum_pad_tokens = (tokens_per_expert_block_padded - tokens_per_expert).cumsum(0)
+    padded_tokens = torch.zeros(
+        max_total_padded_tokens - topk_ids.numel(),
+        dtype=topk_ids.dtype,
+        device=topk_ids.device,
+    )
+    ones = torch.ones_like(padded_tokens)
+    padded_tokens.scatter_add_(0, cum_pad_tokens[:-1], ones)
+    padded_tokens = padded_tokens.cumsum(0)
+    padded_token_ids_per_block = torch.cat([topk_ids.view(-1), padded_tokens]).argsort()
+
+    return padded_token_ids_per_block, expert_block_mapping, total_padded_tokens
+
+
+lib = torch.library.Library("moe", "FRAGMENT")
+lib.define(
+    "moe_mm(Tensor input, Tensor moe_matrix, Tensor token_expert_mapping, Tensor padded_token_ids_per_block, Tensor expert_block_mapping, Tensor total_padded_tokens, int topk, int padding_size) -> Tensor"
+)
+
+
+# All that's needed for torch.compile support
+@torch.library.impl(lib, "moe_mm", "Meta")
```

Comment:
not really relevant to this particular fix but are we supposed to use torch.library.impl_abstract for meta tensors? I'm assuming that's what meta means here

Reply:
yes, although even `impl_abstract` is getting deprecated in 2.4 for a new API, so I'll make the change and when we upgrade to 2.4 we can move to the new one